Allow arbitrary docstrings to be inside torchscript interface methods

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50271 Allow arbitrary docstrings to be inside torchscript interface methods. This bug was introduced due to python 3.8 because of ast library change**

Summary:

Test Plan: new python test case

Reviewers: @gmagogsfm

Subscribers:

Tasks: https://github.com/pytorch/pytorch/issues/50188

Tags:

Differential Revision: [D25853916](https://our.internmc.facebook.com/intern/diff/D25853916)

tugsbayasgalan committed:
Allow arbitrary docstrings to be inside torchscript interface methods

Summary:

Test Plan: new python test case

Reviewers: gmagogsfm

Subscribers:

Tasks: https://github.com/pytorch/pytorch/issues/50188

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Allow arbitrary docstrings to be inside torchscript interface methods"


Summary:

Test Plan: new python test case

Reviewers: @gmagogsfm

Subscribers:

Tasks: https://github.com/pytorch/pytorch/issues/50188

Tags:

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -4323,11 +4324,38 @@ void CompilationUnit::define_interface(
+
+    auto stmts_size = method_def.statements().size();
+    // if there is only one element, it should be "pass" statement
+    if (stmts_size == 1 && method_def.statements()[0].kind() != TK_PASS) {
```
If seems this loop handles only a special case of the following loop? maybe it can be omitted.

tugsbayasgalan replied:
oh true. will fix it

gmagogsfm commented on test/jit/test_module_interface.py:
```diff
@@ -147,6 +147,33 @@ def call_module_interface_on_other_method(mod_interface, x):
+
+        scripted_mod = torch.jit.script(TestModule())
+        input = torch.randn(3, 4)
+        self.assertEqual(scripted_mod(input), 3 * input + 2)
```
Nit: invoke unscripted TestModule to calculate correct result rather than hard coding as 3 * input + 2

gmagogsfm reviewed (approved):
Change looks good, just a few minor comments.

On the high level though, did you confirm if python 3.8 introduced this change? If so, it would be great to add more information on the PR description.

tugsbayasgalan committed:
Update on "Allow arbitrary docstrings to be inside torchscript interface methods"


Summary:

Test Plan: new python test case

Reviewers: @gmagogsfm

Subscribers:

Tasks: https://github.com/pytorch/pytorch/issues/50188

Tags:

[ghstack-poisoned]
